### ISSUE TITLE
Question about project icons

### ISSUE THREAD

**Issue by fgsalomon** [jenkinsci/bitbucket-branch-source-plugin#183]
### Your checklist for this issue

- [x] Jenkins version

- [x] Plugin version

- [ ] Bitbucket cloud

- [x] Bitbucket server and version

### Description

Hi,

We are using Bitbucket Branch Source Plugin version 2.4.4 and everything works great. However, we'd like to know if there is a way of setting projects icons to their Bitbucket avatars.
Right now the projects have random icons:
![imagen](https://user-images.githubusercontent.com/3149695/55898351-afa32b00-5bc2-11e9-808f-388de597423e.png)

Projects avatars are available through Bitbucket REST API so it would be great if the plugin could use them or at least have a way to set them in the configuration.

**Jenkins:** 2.164.2
**Bitbucket Branch Source Plugin:** 2.4.4
**Bitbucket Server:** 5.16.2

Thanks.

**Comment by casz** [comment]
These are already fetched from the API as I recall. Are we talking bitbucket server or bitbucket cloud?

**Comment by lelandsindttouchnet** [comment]
I see the same behaviour. 

We are running Bitbucket Server (6.0.0) and Bitbucket Branch Source Plugin (2.4.4)

... would be glad to provide logs if it will help.

**Comment by fgsalomon** [comment]
> These are already fetched from the API as I recall. Are we talking bitbucket server or bitbucket cloud?

Sorry, forgot to add that. We're using Bitbucket Server (v5.16.2).

**Comment by fgsalomon** [comment]
I've found a bug report [here](https://issues.jenkins-ci.org/browse/JENKINS-45852) that says that this happens since version 2.2.2.

**Comment by ViceIce** [comment]
Any progress here?

**Comment by casz** [comment]
It is up for grabs

**Comment by mlasevich** [comment]
> It is up for grabs

Does that mean it's abandoned? Seems like it has been several years with no progress

**Comment by casz** [comment]
Not sure how it's been several years when this issue was posted this year.

In any case

I am only a maintainer, I don't actively use bitbucket. If someone wants to actively maintain the Bitbucket integration I'd gladly accept a maintainer who actively uses Bitbucket. I'll also gladly review any contributions.

**Comment by mlasevich** [comment]
Makes sense. Hard to really fix anything if you are not actually using it. You are just keeping the lights on. (and much gratitude for that, but I understand why fix for this is not likely in this scenario *sigh* )

It has been a couple of years since the original issue was filed in July of 2017 with JIRA - though I am not sure if filing it with Jenkins Jira means anything in this case (https://issues.jenkins-ci.org/browse/JENKINS-45852)

I am happy to take a stab at providing a patch, but not being familiar with this at all it may take a while. I did have a glance at the code, and from what I saw it seems to be doing the right thing and the actual avatar getting/caching is handled by Jenkins code, not this plugin - but there is no logging, so it is very hard to debug without having to recompile it. For what its worth, I did check the API url for the avatar, and it appears correct.

**Comment by mlasevich** [comment]
Well, I spent some time on it this morning, and I think I know what is happening, though I am not yet sure how to fix it, as I do not know how avatar caching system works...

Short of it, it generates a correct URL for the avatar, and requests it... without credentials. It then gets a 401 and gives up. I am guessing it just passes the URL to avatar caching system and that is what is trying to actually get the image, but it is just a guess at this point... As I know very little about inner workings of Jenkins, I am poking in the dark, so any pointers would be appreciated

**Comment by lelandsindttouchnet** [comment]
maybe not helpful... 

... this explains why the one icon that shows up in Jenkins from Bitbucket in my environment is a Bitbucket project with public repos.

Could it be as simple as passing Bitbucket credentials in the avatar request?

**Comment by ViceIce** [comment]
Some time ago i heard someone said something about that. Public project avatars works but private not.

Some cause could be that the avatar caching system does not now anything about credentials.

**Comment by casz** [comment]
GitLab branch source implemented it's own avatar cache: https://github.com/jenkinsci/gitlab-branch-source-plugin/blob/03acdddf23015ffe679bb61dc6545f0de92ef427/src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabAvatarCache.java

The solution as I see would be implement auth in the SCM API avatar cache: https://github.com/jenkinsci/scm-api-plugin/blob/dba4248724cd53e149186cbf0df063d50bd042ce/src/main/java/jenkins/scm/impl/avatars/AvatarCache.java

As it would benefit all. Though you could go the copy the code and implement auth here but 🤷‍♂

**Comment by mlasevich** [comment]
Thanks. I am looking at Gitlab code and while they did copy the AvatarCache, I see no significant changes from the real one other than some formatting changes and some methods being made public.  I.e. I do not see how they authenticate either - if they authenticate at all :-/ 

For what its worth, I agree with you on adding this functionality to avatar-cache, but I am not sure I know enough to do this right, yet. To start with I opened a [ticket with them ](https://issues.jenkins-ci.org/browse/JENKINS-59797) to see what they think about it...

**Comment by ViceIce** [comment]
I think the fasted solution is to copy the `AvatarCache` and pass our `Credentials` to it, so the fetch can add a `Authorization` header.

Best solution would be to add credentials handling to the scm-api avatar handling.

**Comment by casz** [comment]
I would welcome a PR.

**Comment by mlasevich** [comment]
I'll take a stab at it. Is there a good place to ask questions about plugin development? Slack or something? It took a few minutes, but I already ran into trouble. Avatar-Cache requires features added to Jenkins in a version later than this plugin uses. I bumped version in pom and instantly got an avalanche of enforcer violations... I bypassed enforcer for now, but there is probably a proper way to deal with it :-D

**Comment by casz** [comment]
I can do some TLC to get the project modernised

**Comment by casz** [comment]
Here is a general guide https://jenkins.io/doc/developer/plugin-development/

There are several options for Q&A

https://gitter.im/jenkinsci/jenkins
https://groups.google.com/forum/#!forum/jenkinsci-dev
https://webchat.freenode.net/?channels=jenkins

**Comment by mlasevich** [comment]
Of course the second I sat down to do this, work got in the way, and it took me a while to get back to this. I finally had some time today, so I threw together a basic update to AvatarCache that supports custom sources for avatars (if it works, I will submit it to scm-api-plugin) and switched this plugin to use it with a custom source for BitBucket. I ended up adding retrieving of Avatar to the API object. The good news is that it seems to work, though I only tested lightly and only BB Server - need to test the BB Cloud as well. I ran into the fact that it seems that cloud/server have two completely different implementations - which was surprising...

I was going to submit a PR, but ran into another annoyance. My IDE is set to reformat the files on save, which makes for a really hard to read PR. I am going to go back and undo the damage and then submit PR for prelim review - I just want to make sure I am heading in the right direction :-)
